mesonbuild/mtest.py

```diff
@@ -1849,36 +1863,49 @@ def test_suitable(self, test: TestSerialisation) -> bool:
 
         return True
 
-    def tests_from_args(self, tests: T.List[TestSerialisation]) -> T.Generator[TestSerialisation, None, None]:
+    @staticmethod
+    def make_selection_checker(patterns: T.List[str], should_match_if_no_patterns: bool = False) -> T.Callable[[TestSerialisation], bool]:
         '''
-        Allow specifying test names like "meson test foo1 foo2", where test('foo1', ...)
+        Allow specifying test names like "foo1", which will match test('foo1', ...)
 
         Also support specifying the subproject to run tests from like
-        "meson test subproj:" (all tests inside subproj) or "meson test subproj:foo1"
-        to run foo1 inside subproj. Coincidentally also "meson test :foo1" to
+        "subproj:" (all tests inside subproj) or "subproj:foo1"
+        to run foo1 inside subproj. Coincidentally also ":foo1" to
         run all tests with that name across all subprojects, which is
-        identical to "meson test foo1"
+        identical to "foo1"
         '''
-        for arg in self.options.args:
-            if ':' in arg:
-                subproj, name = arg.split(':', maxsplit=1)
+        if not patterns:
+            return lambda _: should_match_if_no_patterns
+
+        unqualified: T.Set[str] = set()
+        qualified: T.Dict[str, T.Union[T.Set[str], UniversalSet[str]]] = defaultdict(set)
+
+        for pattern in patterns:
+            if ':' in pattern:
+                subproj, name = pattern.split(':', maxsplit=1)
             else:
-                subproj, name = '', arg
-            for t in tests:
-                if subproj and t.project_name != subproj:
-                    continue
-                if name and t.name != name:
-                    continue
-                yield t
+                subproj, name = '', pattern
+
+            if subproj == '':
+                unqualified.add(name)
+            elif name == '':
+                qualified[subproj] = UniversalSet()
```

Comment:
What happens if someone passes in `--exclude ":"`?

Reply:
I think that would be handled by the first branch and not match or exclude anything (tests cannot have empty name).

However, as mentioned above, I do not think a simple rebase is possible since  the set-based approach is not compatible with wildcards.

Reply:
Ah...

I think it's acceptable if excludes don't support wildcards. We can document this and suggest that anyone who wants wildcard support should submit a PR to rewrite the internal implementation details such that wildcards are supported.

My expectation is that in practice, this is only going to be used by people who want to disable a small handful of exact test names, typically because they are e.g. using a distro build recipe and want to skip a test that is known to be failing, but equally known to not be a blocker for updating to new versions.

Reply:
As maintainer for lots of packages in Alpine Linux, I agree that this is useful without wildcard support. If needed it can always be added later.